File: tests/fixtures/corpus-full/ruby/services.rb
require_relative 'models'

class Database
  def initialize
    @store = {}
  end

  def save_user(user)
    @store["user:#{user.id}"] = user
  end

  def find_user(id)
    @store["user:#{id}"]
  end

  def save_product(product)
    @store["product:#{product.id}"] = product
  end

  def find_product(id)
    @store["product:#{id}"]
  end
end

class UserService
  def initialize(db)
    @db = db
  end

  def register(username, email)
    user = User.new(rand(1_000_000), username, email)
    @db.save_user(user)
    user
  end

  def find_by_id(id)
    @db.find_user(id)
  end

  def deactivate(id)
    user = @db.find_user(id)
    return false unless user
    user.deactivate
    @db.save_user(user)
    true
  end
end

class ProductService
  def initialize(db)
    @db = db
  end

  def get_by_id(id)
    @db.find_product(id)
  end

  def restock(id, qty)
    product = @db.find_product(id)
    return false unless product
    product.stock += qty
    @db.save_product(product)
    true
  end
end

File: tests/fixtures/corpus-full/kotlin/Services.kt
package com.example.app

class Database {
    private val store = mutableMapOf<String, Any>()

    fun saveUser(user: User) { store["user:${user.id}"] = user }
    fun findUser(id: Long): User? = store["user:$id"] as? User
    fun saveProduct(product: Product) { store["product:${product.id}"] = product }
    fun findProduct(id: Long): Product? = store["product:$id"] as? Product
}

class UserService(private val db: Database) {
    fun register(username: String, email: String): User {
        val user = User(id = System.currentTimeMillis(), username = username, email = email)
        db.saveUser(user)
        return user
    }

    fun findById(id: Long): User? = db.findUser(id)

    fun deactivate(id: Long): Boolean {
        val user = db.findUser(id) ?: return false
        user.deactivate()
        db.saveUser(user)
        return true
    }
}

class ProductService(private val db: Database) {
    fun getById(id: Long): Product? = db.findProduct(id)

    fun restock(id: Long, qty: Int): Boolean {
        val product = db.findProduct(id) ?: return false
        product.stock += qty
        db.saveProduct(product)
        return true
    }
}

fun main() {
    val db = Database()
    val userService = UserService(db)
    val user = userService.register("alice", "alice@example.com")
    println("Registered: ${user.displayName()}")
}
